presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -234,6 +234,17 @@ private boolean tryFetchLookupSourceProvider()
             }
             lookupSourceProvider = requireNonNull(getDone(lookupSourceProviderFuture));
             statisticsCounter.updateLookupSourcePositions(lookupSourceProvider.withLease(lookupSourceLease -> lookupSourceLease.getLookupSource().getJoinPositionCount()));
+
+            // check if we can finish the join earlier
+            if (!probeOnOuterSide &&
+                    !finishing &&
+                    !finished &&
```

Comment:
Can `finishing == true` at this point?
Can `finished == true` at this point?

Reply:
@findepi `tryFetchLookupSourceProvider` will be called by `getOutput`, and this could be called a few times.

Reply:
I think second call won't reach here, since `this.lookupSourceProvider` will be already set and not null

Reply:
I thought the same thing, however, during my local testing using the `TpchQueryRunner` and running one of the queries I've added to ATQ: I saw an issue with an exception from a `checkState(!finishing)` I had added at some point, based on that I added a conditional break point to output the stack trace if we reached line [235](https://github.com/prestodb/presto/blob/b12002905134caa27570097745fb94d87b30df18/presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java#L235). We do reach that part of the code, it took me a while to reproduce the condition (maybe 1 out of 50 queries):

...
```
[java.lang.Thread.getStackTrace(Thread.java:1559),com.facebook.presto.operator.LookupJoinOperator.tryFetchLookupSourceProvider(LookupJoinOperator.java:235),com.facebook.presto.operator.LookupJoinOperator.addInput(LookupJoinOperator.java:204),com.facebook.presto.operator.Driver.processInternal(Driver.java:385),com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:278),com.facebook.presto.operator.Driver.tryWithLock(Driver.java:645),com.facebook.presto.operator.Driver.processFor(Driver.java:272),com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975),com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162),com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492),java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142),java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617),java.lang.Thread.run(Thread.java:748)]
[java.lang.Thread.getStackTrace(Thread.java:1559),com.facebook.presto.operator.LookupJoinOperator.tryFetchLookupSourceProvider(LookupJoinOperator.java:235),com.facebook.presto.operator.LookupJoinOperator.addInput(LookupJoinOperator.java:204),com.facebook.presto.operator.Driver.processInternal(Driver.java:385),com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:278),com.facebook.presto.operator.Driver.tryWithLock(Driver.java:645),com.facebook.presto.operator.Driver.processFor(Driver.java:272),com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975),com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162),com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492),java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142),java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617),java.lang.Thread.run(Thread.java:748)]
[java.lang.Thread.getStackTrace(Thread.java:1559),com.facebook.presto.operator.LookupJoinOperator.tryFetchLookupSourceProvider(LookupJoinOperator.java:235),com.facebook.presto.operator.LookupJoinOperator.getOutput(LookupJoinOperator.java:295),com.facebook.presto.operator.Driver.processInternal(Driver.java:379),com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:278),com.facebook.presto.operator.Driver.tryWithLock(Driver.java:645),com.facebook.presto.operator.Driver.processFor(Driver.java:272),com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975),com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162),com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492),java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142),java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617),java.lang.Thread.run(Thread.java:748)]
[java.lang.Thread.getStackTrace(Thread.java:1559),com.facebook.presto.operator.LookupJoinOperator.tryFetchLookupSourceProvider(LookupJoinOperator.java:235),com.facebook.presto.operator.LookupJoinOperator.getOutput(LookupJoinOperator.java:295),com.facebook.presto.operator.Driver.processInternal(Driver.java:379),com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:278),com.facebook.presto.operator.Driver.tryWithLock(Driver.java:645),com.facebook.presto.operator.Driver.processFor(Driver.java:272),com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975),com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162),com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492),java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142),java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617),java.lang.Thread.run(Thread.java:748)]
```
...

Reply:
@findepi Please let me know if there is anything else I can address as part of this PR.

Reply:
I don't understand this. Can you please tell me how can i reproduce this? Which exact query have you been running? which test class were you using?

Reply:
I see `finishing=true` is completely legal, e.g. when probe side turns out to be empty too. (please add a test case)

i still don't see how `finished=true` is possible. _Anyway_, checking `&& !finishing`, `&& !finished` seems redundant. What about simply removing those conditions? 

.. Especially that you set both flags. What if `finishing=true` but `finished=false`? The code is still going to work, but if you assume such situation is perfectly OK, then you shouldn't set both flags to true below.

Reply:
@findepi I'll remove the conditions and add the test case. Thanks for taking a look a this.

Reply:
I've addressed all the comments plus added a few test cases, I had already a UT for an empty probe, but not one with build data and an empty probe (added 2 more cases there).